### PR TITLE
Fix trans for foreign methods

### PR DESCRIPTION
**Pull request by nagisa**

### CODE REVIEW

File: src/librustc_trans/trans/base.rs [hunk 1]
```diff
@@ -2423,12 +2423,14 @@ fn register_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
                          node_id: ast::NodeId,
                          node_type: Ty<'tcx>)
                          -> ValueRef {
-    match node_type.sty {
-        ty::ty_bare_fn(_, ref f) => {
-            assert!(f.abi == Rust || f.abi == RustCall);
+    if let ty::ty_bare_fn(_, ref f) = node_type.sty {
+        if f.abi != Rust && f.abi != RustCall {
+            ccx.sess().span_fatal(sp, "this function may only have \"Rust\" or \"rust-call\" \
```
Comment (nikomatsakis) [inline src/librustc_trans/trans/base.rs]: This should be a `span_bug`, not `span_fatal`, unless you think this is user error (the assert suggests it's a bug in the compiler).

Also, it's always nice to print out the value of `f.abi` too.

**Comment by nikomatsakis** [comment]
r+ if the fatal is changed to a bug (ping me or someone else once that's done...)

**Comment by nagisa** [comment]
note to self:
> nmatsakis: it seems like we should report an error in librustc_typeck::astconv::ty_of_method

**Comment by nagisa** [comment]
I updated the code to report error in typeck, and indeed, it is much nicer. 

I’m not happy about the span, though. Currently it spans the whole function and I’d like to only span the `extern( ".*")?` bit, but I can’t see `extern fn` being represented in AST, so I conclude it is not easily doable.

**Comment by nagisa** [comment]
Rebased.

**Comment by Manishearth** [comment]
Needs https://github.com/Manishearth/rust/commit/de6ba0a6599860a73b835cd9956eda62de50dbab to work on master.

**Comment by Manishearth** [comment]
Also https://github.com/rust-lang/rust/commit/315f2002df83c813b4374f19cec0cfa8885f0405 (Windows specific stuff)

**Comment by Aatch** [comment]
Is it possible to get this reverted? Or more accurately, changed to completely disallow non-Rust ABI methods? There's a bunch of bugs related to this and I'm not sure non-Rust ABI methods really make sense.

All this PR did was allow them to be declared, but didn't handle the bodies. It also never addressed methods on trait objects.

**Pull request merged**